src/Runner.Listener/Runner.cs

```diff
@@ -396,7 +396,14 @@ private async Task<int> RunAsync(RunnerSettings settings, bool runOnce = false)
                                     autoUpdateInProgress = true;
                                     var runnerUpdateMessage = JsonUtility.FromString<AgentRefreshMessage>(message.Body);
                                     var selfUpdater = HostContext.GetService<ISelfUpdater>();
-                                    selfUpdateTask = selfUpdater.SelfUpdate(runnerUpdateMessage, jobDispatcher, !runOnce && HostContext.StartupType != StartupType.Service, HostContext.RunnerShutdownToken);
+
+#if OS_WINDOWS
+                                    var restartInteractiveRunner = false;
```

Comment:
is this going to break runners running interactively? also is this going to break runners running as a windows service?

Reply:
i.e. will they come back online

Reply:
It did not affect interactive runners in my tests. This line just disables restarting the runner from the update script:

https://github.com/actions/runner/blob/93ec16e14f6d3bca65df10e137dc0132aed8fe73/src/Misc/layoutbin/update.cmd.template#L128-L133

The runner listener will still return one of these two update related exit codes (4 and 3 respectively):

https://github.com/actions/runner/blob/93ec16e14f6d3bca65df10e137dc0132aed8fe73/src/Runner.Listener/Runner.cs#L352-L359

Both those exit codes are then handled by the run.cmd script. The only difference from a user perspective is that the updated runner reuses the original command prompt window instead of creating a new one.

Reply:
If you run `whoami` to check the runner run-as user, what is the user before and after the update?

Reply:
ok i see now there are two different places where we restart - sometimes the cmd file starts the runner again and sometimes the update script restarts the runner

Reply:
I am not sure whether this is the right fix, 
I think we should check:
- The way we invoke the generated update.cmd script from the runner, we use `Process.Start()`, do we miss anything?
- How the update.cmd restart the runner process, we just `START`, do we miss any parameter?

Reply:
Interesting, I just ran another test where I changed `"%~dp0run.cmd" %*` to `start "Actions Runner" cmd.exe /k "%~dp0run.cmd" %*` in run.cmd, which also resulted in a failure. So `START` certainly compounds the issue, but removing it is not enough. Creating a new process from C# is still an issue.

Reply:
For completeness, I ran another test where I used `invokeScript.StartInfo.UseShellExecute = true` and removed the `START` call from the update script (I wasn't 100% sure I had tested this scenario before). Unfortunately it still resulted in failure.

Reply:
Thanks @MatisseHack this is awesome data

Would you be interested in reopening this PR against a [feature branch](https://github.com/actions/runner/tree/features/consolidate-restart)? I like the proposed approach in this PR, but i think there is some more follow-up we want to do (@fhammerl has bandwidth to follow-through if that helps). I like that we're converging the restart code into one place. We should also delete the code from the update.template.cmd. And we should change Linux/macOS code paths so behaviors align across all platforms.

Since we have additional changes to make, I would rather do one big test pass at the end rather. So i'm hesitant for this to go into main yet. Runner update high risk. If we mess something up, then all customer's self-hosted runners will fail to update or they may have to manually restart their runners. IIRC sometimes hosted runners also encounter the auto-update path too.

@fhammerl one thing to look out for, is to think-through or verify the runner can successfully update twice :) - we had that bug in the past :(

Reply:
No problem, I'll retarget this PR to that branch!

I think keeping Linux and macOS consistent with Windows is a great idea. I'm happy to make those changes if it would help move things along, but happy to let @fhammerl take over as well if you would like.

If there are going to be more involved changes, I also wonder if it might make sense to call (and wait for) the update script from the run script instead of starting it from C# and waiting a set period of time before restarting the runner? Obviously the current approach has been working well so far, but removing the `timeout` would be a small chance to make the process a bit more deterministic.

Also, in regards to removing code from the update script template, I think that will have to wait for a future update. Existing runners expect the update script to restart the runner once it completes. If we remove that code now, runners will never restart after updating because their existing run.cmd script will not have this updated logic yet. Unfortunately we need to get these changes in run.cmd deployed before removing that dead code in the update script.